text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,553 @@
+```
+
+
+# Internal details
+
+![drawing](img/context_propagation_details.png)
+
+## Default HTTP headers
+OpenTelemetry currently uses two standard header formats for context propagation. 
+Their properties and requirements are integrated into the OpenTelemetry APIs.
+
+**Span Context -** The OpenTelemetry Span API is modeled on the `traceparent` 
+and `tracestate` headers defined in the [W3C Trace Context specification](https://www.w3.org/TR/trace-context/). 
+
+**Correlation Context -** The OpenTelemetry Correlations API is modeled on the 
+`Correlation-Context` headers defined in the [W3C Correlation Context specification](https://w3c.github.io/correlation-context/). 
+
+## Context management and in-process propagation
+
+In order for Context to function, it must always remain bound to the execution 
+of code it represents. By default, this means that the programmer must pass a 
+Context down the call stack as a function parameter. However, many languages 
+provide automated context management facilities, such as thread locals. 
+OpenTelemetry should leverage these facilities when available, in order to 
+provide automatic context management.
+
+## Pre-existing context implementations
+
+In some languages, a single, widely used context implementation exists. In other 
+languages, there many be too many implementations, or none at all. For example, 
+Go has a the `context.Context` object, and widespread conventions for how to 
+pass it down the call stack. Java has MDC, along with several other context 
+implementations, but none are so widely used that their presence can be 
+guranteed or assumed.
+
+In the cases where an extremely clear, pre-existing option is not available, 
+OpenTelemetry should provide its own context implementation.
+
+
+# FAQ
+
+## What about complex propagation behavior?
+
+Some OpenTelemetry proposals have called for more complex propagation behavior. 
+For example, falling back to extracting B3 headers if W3C Trace-Context headers 
+are not found. "Fallback propagators" and other complex behavior can be modeled as 
+implementation details behind the Propagator interface. Therefore, the 
+propagation system itself does not need to provide an mechanism for chaining 
+together propagators or other additional facilities.
+
+
+## Did you add a context parameter to every API call because Go has infected your brain?
+
+No. The concept of an explicit context is fundamental to a model where 
+independent cross-cutting concerns share the same context propagation layer. 
+How this context appears or is expressed is language specific, but it must be 
+present in some form.
+
+# Prior art and alternatives
+
+Prior art:  
+* OpenTelemetry distributed context
+* OpenCensus propagators
+* OpenTracing spans
+* gRPC context
+
+# Open questions
+
+Related work on HTTP propagators has not been completed yet.
+
+* [W3C Trace-Context](https://www.w3.org/TR/trace-context/) candidate is not yet 
+  accepted.
```

Comment:
this is state, and also why is trace-context spec relevant for this OTEP?

Reply:
True, this is a distraction. I removed this reference, and cleaned up this section to focus on the fact that the Correlation-Context spec could theoretically diverge from what we intended in this API.